### PR TITLE
Fixes #24676 - Rails 5.2 compatibility

### PR DESCRIPTION
**Pull request by ik5**
connected to https://github.com/theforeman/foreman/pull/5836

### CODE REVIEW

File: app/models/katello/concerns/setting_extensions.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@ module SettingExtensions
       end
 
       def recalculate_errata_status
-        ForemanTasks.async_task(Actions::Katello::Host::RecalculateErrataStatus) if value_changed? && name == 'errata_status_installable'
+        ForemanTasks.async_task(Actions::Katello::Host::RecalculateErrataStatus) if value_previously_changed? || value_changed? && name == 'errata_status_installable'
```
Comment (iNecas) [inline app/models/katello/concerns/setting_extensions.rb]: Parenthesis must be added for this still to work properly:

```ruby
(value_previously_changed? || value_changed?) && name == 'errata_status_installable'
```

Comment (ik5) [reply app/models/katello/concerns/setting_extensions.rb]: yes, fixed that haven't committed it yet

**Comment by iNecas** [comment]
rebase needed

**Comment by iNecas** [comment]
@ik5 I've heard your tests are passing locally… mind pushing it here? :)

**Comment by iNecas** [comment]
Also, let's create issues in Redmine for those that you're addressing here and reference them from the commit, so that we can get closer to actually merging this stuff

**Comment by ik5** [comment]
@iNecas I wanted to keep it for myself, so it will be all mine, but you requested so nicely, so here you have my precious (tests) :P

**Comment by ik5** [comment]
[Created support](https://projects.theforeman.org/issues/24676), now need to change everything to support this bug report, but following conversation with @tbrisker, I should not 
squash the commits.

**Comment by iNecas** [comment]
I think there is not that much changes in this PR that would not handle squashing. We can keep adding new commits as separate changes if needed.

**Comment by iNecas** [comment]
Also, I guess, we can remove the WIP part.

File: app/models/katello/concerns/host_managed_extensions.rb [hunk 2]
```diff
@@ -82,8 +82,19 @@ def correct_puppet_environment
...
+        # IMPORTANT:
+        #
+        # Since rails 5.2, _before_last_save and _was exists on different stages
+        # of callback cycles the _was exists on "before_save", and
+        # "_before_last_save" "after_save" here the following method is called
+        # on both callback hooks, so it is validating both cases.
+        #
+        # Additional important thing to know, is that "_was" changes it's value
+        # after "save" was called, and it is pointing to the current value of
+        # that field, so just using the "_was" will trigger it's own issues.
```
Comment (jlsherrill) [inline app/models/katello/concerns/host_managed_extensions.rb]: This style of comment seems more appropriate as a commit message rather than a comment in the code, do you mind moving it to the commit?

Comment (jlsherrill) [reply app/models/katello/concerns/host_managed_extensions.rb]: Everything else looks great and already covered by unit tests!

**Comment by jlsherrill** [comment]
[test katello]

**Comment by jlsherrill** [comment]
[test katello]

File: app/models/katello/concerns/host_managed_extensions.rb [hunk 3]
```diff
@@ -82,8 +82,19 @@ def correct_puppet_environment
...
+        content_facet &&
+          (content_facet.content_view_id_before_last_save || content_facet.content_view_id_was) == environment.try(:content_view).try(:id) &&
```
Comment (tbrisker) [inline app/models/katello/concerns/host_managed_extensions.rb]: if the previous value was `nil` and this is called from `after_save`, we will get an unexpected value - we want to know what the original value was, but we'll get the new value instead.

**Comment by iNecas** [comment]
@ik5 mind rebasing so that we can run a jenkins job in combination with https://github.com/theforeman/foreman/pull/5836 and see how the jenkins looks like with that. The rebase is needed to get https://github.com/Katello/katello/pull/7639

File: app/models/katello/concerns/location_extensions.rb [hunk 4]
```diff
@@ -15,11 +15,11 @@ def set_default_overrides
       end
 
       def reset_settings
```
Comment (tbrisker) [inline app/models/katello/concerns/location_extensions.rb]: This is only called from `after_save`, no need to handle the `before_save` behavior here (the new functions are already present in 5.1).

File: app/models/katello/concerns/setting_extensions.rb [hunk 5]
```diff
@@ -11,7 +11,7 @@ module SettingExtensions
       end
 
       def recalculate_errata_status
-        ForemanTasks.async_task(Actions::Katello::Host::RecalculateErrataStatus) if saved_change_to_value? && name == 'errata_status_installable'
+        ForemanTasks.async_task(Actions::Katello::Host::RecalculateErrataStatus) if (value_previously_changed? || value_changed?) && name == 'errata_status_installable'
```
Comment (tbrisker) [inline app/models/katello/concerns/setting_extensions.rb]: This is also only called in `after_save`

File: app/models/katello/concerns/host_managed_extensions.rb [hunk 6]
```diff
@@ -82,8 +82,17 @@ def correct_puppet_environment
       end
 
       def content_and_puppet_matched?
```
Comment (tbrisker) [inline app/models/katello/concerns/host_managed_extensions.rb]: This is only called from `correct_puppet_environment` which is called in `before_save`, so this change shouldn't be needed.

File: app/models/katello/concerns/location_extensions.rb (hunk 4 above)
Comment (ik5) [reply app/models/katello/concerns/location_extensions.rb]: all of the code that have both cases, are called, not always directly from both hooks, i didn't wrote it just to cover possibilities, but only on cases where it happened

**Comment by ik5** [comment]
@tbrisker  all of the code that have both cases, are called, not always directly from both hooks, i didn't wrote it just to cover possibilities, but only on cases where it happened

File: app/models/katello/concerns/location_extensions.rb (hunk 4 above)
Comment (tbrisker) [reply app/models/katello/concerns/location_extensions.rb]: Where? I can't find any other place in the code where these functions are called:
```
2.5.1 ~/gits/katello (master) $ git grep reset_setting
app/models/katello/concerns/location_extensions.rb:        after_save :reset_settings
app/models/katello/concerns/location_extensions.rb:      def reset_settings
```
and the same is true for the other two.

**Comment by ehelms** [comment]
When test building Katello locally for RPM, I am hitting the following output which comes form running this stage of RPM build:

https://github.com/theforeman/foreman-packaging/blob/rpm/develop/packages/foreman/foreman/foreman.spec#L1145

```
BUILDSTDERR: Caused by:
BUILDSTDERR: SQLite3::SQLException: too many terms in compound SELECT
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/sqlite3-1.3.13/lib/sqlite3/database.rb:91:in `initialize'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/sqlite3-1.3.13/lib/sqlite3/database.rb:91:in `new'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/sqlite3-1.3.13/lib/sqlite3/database.rb:91:in `prepare'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/sqlite3-1.3.13/lib/sqlite3/database.rb:137:in `execute'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/sqlite3_adapter.rb:255:in `block (2 levels) in execute'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/sqlite3_adapter.rb:254:in `block in execute'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:579:in `block (2 levels) in log'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/ruby/monitor.rb:226:in `mon_synchronize'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:578:in `block in log'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract_adapter.rb:569:in `log'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/sqlite3_adapter.rb:253:in `execute'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:1057:in `assume_migrated_upto_version'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/schema.rb:54:in `define'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/schema.rb:46:in `define'
BUILDSTDERR: /usr/share/foreman/schema_plugin.rb:13:in `<top (required)>'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/dependencies.rb:281:in `load'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/dependencies.rb:281:in `block in load'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/dependencies.rb:253:in `load_dependency'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activesupport-5.2.1/lib/active_support/dependencies.rb:281:in `load'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/tasks/database_tasks.rb:245:in `load_schema'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/tasks/database_tasks.rb:266:in `block in load_schema_current'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/tasks/database_tasks.rb:316:in `block in each_current_configuration'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/tasks/database_tasks.rb:313:in `each'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/tasks/database_tasks.rb:313:in `each_current_configuration'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/tasks/database_tasks.rb:265:in `load_schema_current'
BUILDSTDERR: /opt/theforeman/tfm-ror52/root/usr/share/gems/gems/activerecord-5.2.1/lib/active_record/railties/databases.rake:258:in `block (3 levels) in <top (required)>'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/task.rb:251:in `block in execute'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/task.rb:251:in `each'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/task.rb:251:in `execute'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/task.rb:195:in `block in invoke_with_call_chain'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/ruby/monitor.rb:226:in `mon_synchronize'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/task.rb:181:in `invoke'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:160:in `invoke_task'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:116:in `block (2 levels) in top_level'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:116:in `each'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:116:in `block in top_level'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:125:in `run_with_threads'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:110:in `top_level'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:83:in `block in run'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:186:in `standard_exception_handling'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/lib/rake/application.rb:80:in `run'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/exe/rake:27:in `<top (required)>'
BUILDSTDERR: /opt/rh/rh-ruby25/root/usr/bin/rake:23:in `load'
```

**Comment by iNecas** [comment]
Eric: I've tracked the issue to this commit in rails https://github.com/rails/rails/commit/a2827ec9811b5012e8e366011fd44c8eb53fc714#diff-a0775e1ec64264dc76a8ee71a5211ae3R1053

What happens is:

1. in packaging, we first call db:migrate without the plugin enabled, and save the database schema
2. then we enable the plugin and use db:schema:load
3. what rails does inside is, it looks at all the schema versions in the code, and assumes that all the versions earlier than the version of the current schema dump, are migrated (there is not enough info in the schema file to determine, which versions actually were migrated)
4. it inserts the versions into the schema_migrations in bulk

What happens now during phase (4) is failure, because the amount of versions inserted in bulk is just too much for sqlite to handle.

The difference between Rails 5.1 and 5.2 is however actually in the point (3), where there was a bug fixed in Rails. Before the change, it was ignoring the plugin versions when determining, which versions were already migrated. After the change, the katello versions are also scanned, which leads to falsely inserting also versions, that were not actually migrated, and as a side-effect, also crossing the limit of values to be inserted into sqlite at once.

The outcome of this is, until the Rails is fixed, there is a limited amount of versions that can be included during the db:schema:load with sqlite. Even with that fixed however, it would not work for us, as it would assume there were migrations that actually were not applied (because the schema:dump was done before enabling the plugin).

**TL;DR solution:**
What we are doing wrong in packaging is the fact that we call `db:schema:dump` without plugin and `db:schema:load` with plugin. Doing both `db:schema:dump` and `db:schema:load` with plugin enabled would not work, because it would cross the hard limit on amount of migrations. Therefore, the only possible fix right now is doing the `db:schema:load` before we enable the plugin (https://github.com/theforeman/foreman-packaging/blob/rpm/develop/packages/foreman/foreman/foreman.spec#L1137). We would still risk the situation, where we cross the limit of migrations given by the Rails bug, but for now, if would still work, and perhaps, until we'll get there, the Rails will already be fixed.


Alternatively, we can just stop using the `db:schema:load` at all, which was added purely for speeding up the process a bit.

**Comment by ehelms** [comment]
@iNecas Great find and explanation! I'll give changing the ordering a try first to see if that will work for now and report back.

**Comment by ehelms** [comment]
So after some testing I ended up having to drop the schema loading in my testing to get past the current set of issues. Even when I moved plugin activation after schema loading I hit a variety of migration errors. I still do not have successful builds, but now the errors seem to either be Ruby 2.5 or Rails 5.2.1 changes that we need to make?

```
BUILDSTDERR: SyntaxError: /builddir/build/BUILDROOT/tfm-rubygem-katello-3.9.0-3.el7.x86_64/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.9.0/app/controllers/katello/api/v2/repositories_controller.rb:19: syntax error, unexpected '{', expecting keyword_end
BUILDSTDERR: ...n :only => [:create, :update] { find_content_credential CONT...
BUILDSTDERR: ...                              ^
BUILDSTDERR: /builddir/build/BUILDROOT/tfm-rubygem-katello-3.9.0-3.el7.x86_64/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.9.0/app/controllers/katello/api/v2/repositories_controller.rb:20: syntax error, unexpected '{', expecting keyword_end
BUILDSTDERR: ...n :only => [:create, :update] { find_content_credential CONT...
BUILDSTDERR: ...                              ^
BUILDSTDERR: /builddir/build/BUILDROOT/tfm-rubygem-katello-3.9.0-3.el7.x86_64/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.9.0/app/controllers/katello/api/v2/repositories_controller.rb:21: syntax error, unexpected '{', expecting keyword_end
BUILDSTDERR: ...n :only => [:create, :update] { find_content_credential CONT...
BUILDSTDERR: ...                              ^
BUILDSTDERR: /builddir/build/BUILDROOT/tfm-rubygem-katello-3.9.0-3.el7.x86_64/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.9.0/app/controllers/katello/api/v2/repositories_controller.rb:22: syntax error, unexpected '{', expecting keyword_end
BUILDSTDERR: ...n :only => [:create, :update] { find_content_credential CONT...
```

**Comment by iNecas** [comment]
@ehelms this for sure is Ruby 2.4 -> 2.5 thing: This used to work in 2.4:

```ruby
hello :hello => "world' { puts 'hello' }
```

The fix is either using extra parenthensis or `do…end`:

```ruby
# this works
hello(:ahoj => [1,2]) { puts 'hello' }

# this also works
hello :ahoj => [1,2] do
  puts 'hello'
end
```
@ehelms is it possible to enable Katello tests for Ruby 2.5?

Anyway, this is not related to this PR and should be handled separately.

**Comment by ehelms** [comment]
I'll look into switching Katello tests to 2.5.

**Comment by ehelms** [comment]
Opened Ruby 2.5 support fix -- https://github.com/Katello/katello/pull/7653

**Comment by tbrisker** [comment]
Looks like a couple of my changes weren't backward compatible with 5.1 due to a change in how `*_previously_changed?` is handled in `after_` callbacks, opened https://github.com/ik5/katello/pull/2 to change the methods to backward compatible versions that will work with both 5.1 and 5.2.

**Comment by ehelms** [comment]
The Ruby 2.5 changes were merged and with this change Katello now builds against the SCL.

**Comment by ehelms** [comment]
I am testing a few things to see about running this against foreman branch in case anything fails between now and then we did have a green build already.

**Comment by ehelms** [comment]
Testing -- [test katello]

**Comment by ehelms** [comment]
Testing -- [test katello]
